### ISSUE TITLE
Universal bounding box regression

### ISSUE THREAD

**Issue by manfprantl** [davisking/dlib#2352]
I am trying to write a universal (i.e. class agnostic) bounding box regression method in Dlib. So given a proposed approximate bounding box the method should be able to refine the bounding box independent of the actual object class (as long as there is some sort of prominent object close to the proposed bbox). 

Would you think that Dlib's shape predictor will be capable to do that or should I rather resort to a DNN approach?

**Comment by arrufat** [comment]
Hi, I think it's possible, even though I've never tried it myself.
In the [dnn_mmod_find_cars_ex](http://dlib.net/dnn_mmod_find_cars_ex.cpp.html) example program you can see that a shape predictor is used for that single class detector. I guess it shouldn't matter that much if boxes are coming from different classes.

I was interested into that kind of bounding box regression for a while, so after digging a bit, I found a [KITTI submission code](http://dlib.net/files/dlib_kitti_submission_mmodCNN_basic7convModel.tar.bz2).
There, you can see how that kind of shape predictor can be trained.

You could also use the [loss_mean_squared_multioutput](http://dlib.net/dlib/dnn/loss_abstract.h.html#loss_mean_squared_multioutput_) to train that kind of bounding box regression if you want to use a DNN approach.

That being seed, I think it's easier to use the built-in bounding box regression in the [`mmod_options`](http://dlib.net/dlib/dnn/loss_abstract.h.html#mmod_options) if you're training a detector from scratch.
``` c++
mmod_options options(target_size, min_target_size);
options.use_bounding_box_regression = true;
options.bbr_lambda =  100;  // default value
```

**Comment by manfprantl** [comment]
Thank you for your reply.
1.)
I already tried a DNN solution with loss_mean_squared_multioutput. It worked reasonably well. However, the accuracy of the bbox regression could have been better. I guess this might have been due to the downsampling of the cnn part of the network before the fc layers or the MSE loss is not as good as a IoU loss (or a variant of it like a DIoU loss). 

2.)
My attempts with the shape predictor were not successful so far. The result was largely influenced by the initial bbox proposal and it did not really correct it to the bbox of the prominent object. Maybe this was due to the training data I used (VOC2012; I noticed that the truth bboxes are not really accurate in many cases) or it is a general limitation of the dlib shape predictor for this problem. I don't know. Any insight would be appreciated. 

3.)
I am not sure if the mmod_detector will work well for this kind of problem (if there is no specific object class with lots of examples but lots of different object classes with just a few examples). Nevertheless, it might be worth trying.

**Comment by davisking** [comment]
It's complicated and it depends on exactly what you mean by "universal".  I
don't really expect the regression tree method to be as good as a CNN at
recognizing a really diverse range of object types.  And yeah, having high
quality training data is a huge deal.  It's often where most of your time
should be spent, since time making better training data often gives much
larger improvements in model accuracy than clever model tweaks.  Not that
clever modeling isn't important too. But people very often take data as a
given, which is a terrible mistake in many cases.

**Comment by manfprantl** [comment]
Hello Davis.

Thank you for taking the time to reply. 
I totally agree with you about the quality of training data. One cannot expect a model to perform well if the input was rubbish. 

Nevertheless, as you said the DNN approach is probably more powerful and this already showed up in my first trials. Even a rather simple regression model (basically a couple of downsampling CNN layers plus some fc layers) worked reasonably well. I just wasn't satisfied with the accuracy so I try to improve that now. 
Probably, I have to combine information from higher resolution layers with lower resolution layers (much like a UNET type) AND I will look for a dataset with better ground truth data. 

Greetings, Manfred

**Comment by arrufat** [comment]
@manfprantl I was curious about how I would use the `loss_mmod` for this task, so last night I trained a detector on a small dataset that I'm building, it has 700 images and 19 classes, and there are around almost 2000 labeled objects. It is very imbalanced, though, with some classes having only 5 samples.

So, I decided to train your "universal" regressor by assigning the same label "object" to all samples and train a single class detector that would detect any object and predict a bounding box for it. Much like the _objectness_ score in the YOLO detectors, but without the classifier part. I ended up with 3 anchor boxes (`object:32x45,58x32,32x79`) and the model is performing reasonably well. Maybe that can be an approach to solve your problem.

**Comment by manfprantl** [comment]
@arrufat I really appreciate the effort you put into this problem. loss_mmod could be another way to solve my problem. Did you use bounding_box_regression=True on your dataset? If yes, how accurate where the predicted bboxes?

As mentioned above, a simple regression network also works reasonably well but I was not satisfied with the accuracy. I will now try a UNET style approach to see if this improves the quality. I will also see if I can find a training dataset with better bboxes than VOC2012.

**Comment by arrufat** [comment]
@manfprantl yes, I enabled the bounding box regression, with the default lambda (100).
Not to brag, but I annotated the bounding boxes myself on that dataset, so they are quite correct and consistent, here's the output of the detector on one image (it's still training...)

![image](https://user-images.githubusercontent.com/1671644/116197390-ae280980-a76f-11eb-8e6a-8993380047b5.png)

As you can see, objects are varied and in shape and size but the bounding boxes are pretty tight.

**Comment by manfprantl** [comment]
@arrufat That looks pretty good. I will probably give it a try.
